Add debugging a ci dump documentation template and script to fill info and include it on azdo test tab (#43937)
dotnet/runtime: eng/testing/debug-dump-template.md

### How do I install it?

You just need to run:

```script
dotnet tool install --global runfo
```

If you already have it installed, make sure you have at least version `0.6.1` installed, which contains support to download helix payloads. If you don't have the latest version just run:

```script
dotnet tool update --global runfo
```

## Download helix payload containing symbols:

You can just achieve this by running:

```script
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o <out-dir>
```

> NOTE: if the helix job is an internal job, you need to pass down a [helix authentication token](https://helix.dot.net/Account/Tokens) using the `--helix-token` argument.

This will download the workitem contents under `<out-dir>\workitems\` and the correlation payload under: `<out-dir>\correlation-payload\`. 

> The correlation payload is usually the testhost or core root, which contain the runtime and dotnet host that we use to run tests.

Once you have those assets, you will need to extract the testhost or core root. Then extract the workitem assets into the same location where coreclr binary is.

## Windows dump on windows

### Debug with WinDbg

> 1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).

Reviewer: With the "internal" windbg sos.dll should be automatically be loaded from the extension gallery. windbg/cdb can be installed from \\dbg\privates\rc or \\dbg\privates\latest so installing dotnet-sos isn't necessary. It is a lot more manual process too, FYI, `dotnet tool install -g dotnet-sos`,  `dotnet-sos install` which displays the command too load this sos.dll like'.load C:\Users\mikem\.dotnet\sos\sos.dll'.

Linux dumps should also work with the latest windbg.

Author: I think we shouldn't add internal instructions here since this can also be done by contributors... but I'm fine if we add a note for that.

Reviewer: Yes, I didn't think about extern contributors. They don't have access to this version of windbg and the extension gallery. So I guess you are on the right track with the more explicit instructions about installing dotnet-sos and explicitly loading sos. Just remember that in the internal windbg will already be already loaded with what should be the same version as dotnet-sos install.

2. Load the dump with a recent WinDbg.
3. Then run the following commands:

```script
!setclrpath <path to core root or testhost where coreclr is>

> !sethostruntime '<path to your local dotnet runtime or testhost>'

Reviewer: If devs have .NET Core installed sos will find it.  !sethostruntime is only necessary if it isn't or installed in a non-standard path.

.sympath+ <directory with symbols (for library tests these are in testhost dir)>
```
### Analyze with dotnet-dump

1. Install [dotnet-dump global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump).
2. Run: `dotnet-dump analyze <path-to-dump>`
3. Then run the following commands:

```script
setclrpath <you should see `Load path for DAC/DBI: '<none>'`>
setclrpath <path to core root or testhost where coreclr is>
setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
```

## Linux dumps on windows

In order to debug a Linux dump on Windows, you will have to first go to the PR/CI build
that sent the test run and download the cross DAC.

Go to your build by navigating to [your build](https://dnceng.visualstudio.com/public/_build/results?buildId=%BUILDID%&view=artifacts&type=publishedArtifacts) and download the `CoreCLRCrossDacArtifacts`, then choose the cross DAC you need depending on the flavor.

Once you've downloaded that artifact, make sure to extract it in the same location where coreclr binary is.

### Debug with WinDbg

> 1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).

Reviewer: Same thing as above.

2. Load the dump with a recent WinDbg.
3. Then run the following commands:

```script

> .load '<path to sos (can get it by running `dotnet-sos install`)>'

Reviewer: If you go with the manual load route, this line should also be in the above script for Windows dumps.

Reviewer 2: And make sure to unload the one that might be bundled first. This is only necessary if the version of SOS in the gallery doesn't have the needed commands and bug fixes yet. This is probably not all too common.

Author: When opening the Linux dump it didn't automatically load the sos plugin for me, but when opening a Windows dump it did... do we know why? I'll have to try this again.

Reviewer 2: On linux, did you run `dotnet sos install`?

Reviewer: I think he means linux dumps under windbg and I think there is a bug that I need to follow up on that sos doesn't get loaded for linux dumps (triggered by the libcoreclr.so module). The command `.update sos` should load it. I need to follow up with the windbg about why it isn't automatically being loaded.

Reviewer: I have version 10.0.20212.1001 and it loads sos.dll for my linux dumps now. Maybe you need a newer windbg version.

Author: > I have version 10.0.20212.1001 and it loads sos.dll for my linux dumps now. Maybe you need a newer windbg version.

Ah that's it. I have a quite old version. Does it make sense mentioning `.update sos` command just in case?

Reviewer: You can mention it, but it would be better to make everybody has a new version.

Author: Sounds good.

!setclrpath <path to core root or testhost where coreclr is>
!sethostruntime '<path to your local dotnet runtime>'
.sympath+ <directory with symbols (for library tests these are in testhost dir)>
```
### Analyze with dotnet-dump

1. Install [dotnet-dump global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump).
2. Run: `dotnet-dump analyze <path-to-dump>`
3. Then run the following commands:

```script
setclrpath <you should see `Load path for DAC/DBI: '<none>'`>
setclrpath <path to core root or testhost where coreclr is>
setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
```

## Linux dumps on Linux

### Debug with LLDB

1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).
2. Load the dump by running `lldb -c <path-to-dmp> <host binary used (found in testhost)>`
3. Run the following commands:

```script
setclrpath <path to core root or testhost where coreclr is>
sethostruntime '<path to your local dotnet runtime or testhost where coreclr is>'
setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>

> # On linux (native symbols are loaded separately on lldb using `target symbols add <symbol file>`)

Reviewer: The `loadsymbols` command should also try to find/download/load the native module's symbols.

Reviewer 2: This is new to me. Thanks Mike. I usually had to load coreclr's separately for some reason.

# Debugging a CI dump

This document describes how to debug a CI/PR test dump by downloading assets from helix, using a dotnet tool called `runfo`.

## What is runfo?

Runfo is a dotnet global tool that helps get information about helix test runs and azure devops builds. For more information see [this](https://github.com/jaredpar/runfo/tree/master/runfo#runfo)

### How do I install it?

You just need to run:

```script
dotnet tool install --global runfo
```

If you already have it installed, make sure you have at least version `0.6.1` installed, which contains support to download helix payloads. If you don't have the latest version just run:

```script
dotnet tool update --global runfo
```

## Download helix payload containing symbols:

You can just achieve this by running:

```script

> runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o <out-dir>

Reviewer 3: Can you outline from where someone gets the relevant jobid and workitem?

Author: This will be filled in by the python script. So the one that will show up by the dump will contain that info there already, that's why I didn't put that info. I was going to put it but then when I so the produced file it didn't make sense. Thoughts on that?

Reviewer 3: You're saying I literally write `-j %JOBID% -w %WORKITEM%` rather than supplying those values myself?  Why do I even need to write that then, i.e. why is the command not just `runfo get-helix-payload -o <out-dir>`?

Author: No not really. What I'm saying is that this is just a template that we feed into the python script that I'm adding as part of this PR and the python script will fill in this information and produce a .md file with all the information in it. For example, this .md file was generated from a crash I forced: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-229762be8cb749d196/Microsoft.CSharp.Tests/debug-dump.md?sv=2019-07-07&se=2020-11-18T23%3A31%3A30Z&sr=c&sp=rl&sig=YzLMjEzJP8W3zhGSp5m2JVzphr%2FgNVm0xRO%2FnV4wNJk%3D

You can see that the command it has is: 
```
runfo get-helix-payload -j 229762be-8cb7-49d1-960a-83e041494e15 -w a760fa7c-e088-41ca-b565-640db87df271 -o <out-dir>
```

So what I was trying to say, is that if you just read the template checked in into the repo, it makes sense to add the information from where to get the jobid... however the purpose of this template is to produce a .md file that is side by side with the dump in the helix logs container, so in the final .md file it will be confusing to have a section to explain "how to get the helix jobid" when you don't even need it since the jobid is already embedded into the .md file you're reading. 

Does that makes it more clear?

Reviewer 3: Sorry, I'm still not understanding.  The directions say all I have to do is:
```
dotnet tool install --global runfo
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o outputdir
```
How does it know what job I care about / what to download?  So is there a step missing from the instructions?

Reviewer 2: This is a template. It gets populated before it the upload to the AzDO tab.

Reviewer 3: Populated with what?  I haven't told it what I care about.

I'm confused.  I want to debug an issue that occurred in CI (which is what this doc is helping me with).  What exactly do I type at the command line?  Let's say my PR for dotnet/runtime PR 12345 failed.  What exactly do I do?

Author: It will be populated with the actual helix job id and workitem id for the crash that you hit in CI.

We populate it here: https://github.com/dotnet/runtime/pull/43937/files#diff-fd8bb0da491f288f6e268f835a49b14126ba3f1a15ab6971ddc76dd480b81b11R95

We pass down that information here: https://github.com/dotnet/runtime/pull/43937/files#diff-1d68a522b60c1ab69fb537ce06f5d7e9a176fa4f75e16425fc164d4b738d7678R216

> What exactly do I type at the command line? Let's say my PR for dotnet/runtime PR 12345 failed. What exactly do I do?

You will go to the azure devops test tab which will show the failure and then in the helix attachments you will see the dump and a MarkDown file. That MD file, will be this template but with the info that you need embedded in there so that you don't have to go and get the helix job yourself. It will look like this:

![image](https://user-images.githubusercontent.com/22899328/97765068-f79fe080-1acd-11eb-9cae-82f91418f382.png)

So the steps you'll follow is download the dump and debug-dump.md file (which I will rename to how-to-debug-dump.md) and just follow the instructions in the MD file, no need to get more information. This is trying to minimize the steps people have to do and info they need to gather whenever they see a dump in their test results. The final MD file will look like [this one](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-229762be8cb749d196/Microsoft.CSharp.Tests/debug-dump.md?sv=2019-07-07&se=2020-11-18T23%3A31%3A30Z&sr=c&sp=rl&sig=YzLMjEzJP8W3zhGSp5m2JVzphr%2FgNVm0xRO%2FnV4wNJk%3D). It will even add a link to the build artifacts in case you need to download the cross DAC to debug a linux dump on Windows. 

> Go to your build by navigating to [your build](https://dnceng.visualstudio.com/public/_build/results?buildId=870703&view=artifacts&type=publishedArtifacts) and download the `CoreCLRCrossDacArtifacts`, then choose the cross DAC you need depending on the flavor.

Reviewer 3: I see.  I totally misunderstood the purpose of this documentation. It's not documentation for someone to read in the dotnet/runtime repo.  It's effectively part of our CI tooling. Thanks :)

Author: > I see. I totally misunderstood the purpose of this documentation. It's not documentation for someone to read in the dotnet/runtime repo. It's effectively part of our CI tooling. Thanks :)

Yeah, that's why I put it under `eng\testing`... however, it can serve as documentation as well, but it will just miss getting this piece of information. I could add it and then strip it out when generating the template if you'd like that?

Reviewer 3: > I could add it and then strip it out when generating the template if you'd like that?

Nah, it's fine, but thanks.

```

> NOTE: if the helix job is an internal job, you need to pass down a [helix authentication token](https://helix.dot.net/Account/Tokens) using the `--helix-token` argument.

This will download the workitem contents under `<out-dir>\workitems\` and the correlation payload under: `<out-dir>\correlation-payload\`. 

> The correlation payload is usually the testhost or core root, which contain the runtime and dotnet host that we use to run tests.

Once you have those assets, you will need to extract the testhost or core root. Then extract the workitem assets into the same location where coreclr binary is.

> ## Windows dump on windows

Reviewer 4: ```suggestion
## Windows dump on Windows
```

### Debug with WinDbg

> 1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).

Reviewer 2: The user still needs to run `dotnet sos install`. Perhaps also worth noting - this has an architecture flag to install different versions of the plugin for scenarios like arm on arm64 and such.

Author: Yeah good suggestion.

> 2. Load the dump with a recent WinDbg version for it to load sos automatically (if not you can run `.update sos`).

Reviewer 3: It's worth noting that the bitness of windbg has to match the bitness of the dump.

Author: Thanks for the suggestion, will include that.

3. Then run the following commands:

```script
!setclrpath <path to core root or testhost where coreclr is>
.sympath+ <directory with symbols (for library tests these are in testhost dir)>
```
### Analyze with dotnet-dump

1. Install [dotnet-dump global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump).
2. Run: `dotnet-dump analyze <path-to-dump>`
3. Then run the following commands:

```script

> setclrpath <you should see `Load path for DAC/DBI: '<none>'`>

Reviewer 2: There's no distinction here between the two `setclrpath`. The first one is to verify a wrong DAC hasn't been loaded. That one has no parameters. The second one should have a path to the folder containing coreclr as the parameter.

setclrpath <path to core root or testhost where coreclr is>
setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
```

> ## Linux dumps on windows

Reviewer 4: ```suggestion
## Linux dumps on Windows
```

setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
```

## Linux dumps on Linux

### Debug with LLDB

1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).
2. Load the dump by running `lldb -c <path-to-dmp> <host binary used (found in testhost)>`
3. Run the following commands:

```script
setclrpath <path to core root or testhost where coreclr is>
sethostruntime '<path to your local dotnet runtime or testhost where coreclr is>'
setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
loadsymbols (if you want to resolve native symbols)
```

### Analyze with dotnet-dump

1. Install [dotnet-dump global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump).
2. Run: `dotnet-dump analyze <path-to-dump>`
3. Then run the following commands:

```script
setclrpath <you should see `Load path for DAC/DBI: '<none>'`>
setclrpath <path to core root or testhost where coreclr is>
setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
```

## MacOS dumps

> MacOS instructions are the same as [Linux](#linux-dumps-on-linux); however MacOS has some caveats.

Reviewer 2: ```suggestion
Instructions for debugging dumps on MacOS the same as [Linux](#linux-dumps-on-linux); however there are a couple of caveats.
```

> 1. It's only supported to debug them in `dotnet-dump` if it's a `createdump` dump.

Reviewer 2: ```suggestion
1. It's only supported to debug them in `dotnet-dump` if it's a runtime generated dump. This includes hang dumps and dumps generated by `createdump` and `dotnet-dump` and the runtime itself. 
```
